### ISSUE TITLE
False positive "nested constraint" exception when using an argument constraint of the wrong type

### ISSUE THREAD

**Issue by progala2** [FakeItEasy/FakeItEasy#1230]
*Original title: **Weird issue with nested constraints and simple types***

```csharp
public interface IEmailFactory
    {
        Task<IList<MailMessage>> GetEmailMessages(long idPairGame, EmailTemplateType emailTemplateType);
    }

_emailFactory = new Fake<IEmailFactory>().FakedObject;

A.CallTo(() => _emailFactory.GetEmailMessages(A<int>._, A<EmailTemplateType>._)).Returns(mailMessages);

```
EmailTemplateType is an enum. When I change it to int it is still wrong, so it is not an issue.

FakeIt version 4.1.

It throws the following exception

>FakeItEasy.Expressions.ExpressionArgumentConstraintFactory.ArgumentConstraintExpressionVisitor.VisitMember(MemberExpression node)
   at System.Linq.Expressions.MemberExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at System.Linq.Expressions.ExpressionVisitor.VisitUnary(UnaryExpression node)
   at System.Linq.Expressions.UnaryExpression.Accept(ExpressionVisitor visitor)
   at FakeItEasy.Expressions.ExpressionArgumentConstraintFactory.CheckArgumentExpressionIsValid(Expression expression)
   at FakeItEasy.Expressions.ExpressionArgumentConstraintFactory.GetArgumentConstraintFromExpression(Expression expression, Object& value)
   at FakeItEasy.Expressions.ExpressionArgumentConstraintFactory.GetArgumentConstraint(ParsedArgumentExpression argument)
   at System.Linq.Utilities.<>c__DisplayClass2_0`3.<CombineSelectors>b__0(TSource x)
   at System.Linq.Enumerable.SelectEnumerableIterator`2.ToArray()
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at FakeItEasy.Expressions.ExpressionCallMatcher..ctor(ParsedCallExpression parsedExpression, ExpressionArgumentConstraintFactory constraintFactory, MethodInfoManager methodInfoManager, Factory outputWriterFactory)
   at FakeItEasy.RootModule.<>c__DisplayClass0_0.<RegisterDependencies>b__4(ParsedCallExpression callSpecification)
   at FakeItEasy.Configuration.FakeConfigurationManager.CallTo[T](Expression`1 callSpecification)

**Comment by progala2** [comment]
It works with 4.0

**Comment by thomaslevesque** [comment]
Thanks for reporting this @progala2. Not sure what's going on. It's a very basic use case, which should be covered extensively by our tests, so I'm really surprised it fails. I'll look into it ASAP.

**Comment by progala2** [comment]
Nope. It works partially with 4.0 - it is registered without errors but doesn't return anything
```csharp
mailsToSend.AddRange(await _emailFactory.GetEmailMessages(game.Id, (EmailTemplateType)mailReminder.EmailTemplate.Id));

Yes it is a very basic case. I tried to clean the solution, I restarted VS... nothing helps, so I post it here :/

**Comment by thomaslevesque** [comment]
There's a problem in your code: it should be `A<long>._`, not `A<int>._`. Since the parameter is of type long, the `A<int>._` constraint will never match; in fact, we even added a diagnostic in the latest [analyzer](https://www.nuget.org/packages/FakeItEasy.Analyzer.CSharp).

I think this is what causing the error, since I can't reproduce the problem when I change the constraint type. However, the exception is clearly incorrect, we definitely have a bug.

**Comment by progala2** [comment]
yes, it works now. Jeez, I hate this type of "typos" 😖 
Thanks 😄

**Comment by blairconrad** [comment]
Not saying this is it, but `idPairGame` is a long but the constraint is `A<int>._`. Consider rationalizing just to rule it out as a problem?

**Comment by blairconrad** [comment]
Bah! Github app didn't refresh the comments. I thought I was so smart.

Glad the problem is found. Definitely a big. And @progala2, do consider picking up the analyzer. It can help in surprising ways.

**Comment by progala2** [comment]
Installed 🗡
Thanks!

**Comment by blairconrad** [comment]
Ha, thank you. And I'm not _just_ pushing it because the diagnostic in question the first one I put any significant effort into... 😉
I hope it helps you.

**Comment by thomaslevesque** [comment]
The problem appears to be [here](https://github.com/FakeItEasy/FakeItEasy/blob/1a3f121f4fcdc7c0a1d356f75178a52023f54005/src/FakeItEasy/Expressions/ExpressionArgumentConstraintFactory.cs#L120-L121). In this case, there is also an implicit conversion, but it's not a boxing conversion, so we don't eliminate it, and the constraint is seen as nested in the conversion. The problem is that there's no way to distinguish between explicit and implicit conversions with the Expression API...

I think we should always eliminate the conversion, even if it's not a boxing conversion, but I'm a bit worried about unwanted side effects.

**Comment by blairconrad** [comment]
@thomaslevesque, this still may not be fixed:

```c#
[Scenario]
public static void ThatArgumentConstraintForDifferentValueTypeWithNonNullArgument(
    IHaveANullableParameter subject,
    int result)
{
    "Given a fake with a method that accepts a nullable value type parameter"
        .x(() => subject = A.Fake<IHaveANullableParameter>());

    "And a call configured for a non-nullable argument of that type"
        .x(() => A.CallTo(() => subject.Bar(A<byte>._)).Returns(42));

    "When I make a call to this method with a non-null argument"
        .x(() => result = subject.Bar(7));

    "Then it matches the configured call"
        .x(() => result.Should().Be(42));
}
```

fails, with a complaint about a nested constraint (the nullable parameter is an `int?`). I think we need to keep stripping conversions, instead of just taking the first.

**Comment by blairconrad** [comment]
Yeah, this seems to fix that portion, at least:

```c#
private static Expression GetExpressionWithoutConversion(Expression expression)
{
    while (expression is UnaryExpression conversion && conversion.NodeType == ExpressionType.Convert)
    {
        expression = conversion.Operand;
    }

    return expression;
}
```

**Comment by thomaslevesque** [comment]
OK, so it's easy enough to fix. Just change an `if` to a `while` ;)

> Of course, this just means we consider the constraint not to be nested. It still does the wrong thing until #1237 is fixed.

True. I have started working on #1237, but it might be a few days before I can send a PR.

**Comment by blairconrad** [comment]
> OK, so it's easy enough to fix. Just change an if to a while ;)

Pretty much. I can work on that later, if you like.

> True. I have started working on #1237

Oh. So have I. It was unowned, so I self-assigned and started working on it last night.

**Comment by thomaslevesque** [comment]
This change has been released as part of [FakeItEasy 4.1.1](https://github.com/FakeItEasy/FakeItEasy/releases/tag/4.1.1).

Thanks for the report @progala2! Look for your name in the release notes. 🏆
